charts/charts.go

```diff
@@ -0,0 +1,162 @@
+package charts
+
+import (
+	"bytes"
+	"errors"
+	"html/template"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/thrasher-corp/gocryptotrader/common/file"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/kline"
+	"github.com/thrasher-corp/gocryptotrader/log"
+)
+
+// New returns a new chart instance
+func New(name, template, outPath string) (chart Chart) {
+	switch template {
+	case "basic":
+		chart.template = "basic.tmpl"
+	case "timeseries":
+		chart.template = "timeseries.tmpl"
+	case "timeseries-markers":
+		chart.template = "timeseries-markers.tmpl"
+	}
+	chart.output = name
+	if outPath == "" {
+		chart.OutputPath = "output"
+	} else {
+		chart.OutputPath = outPath
+	}
+	return chart
+}
+
+// Generate chart output
+func (c *Chart) Generate() (*os.File, error) {
+	var list []string
+	if c.TemplatePath == "" {
+		baseTemplate, err := writeTemplate(templateList["base.tmpl"])
+		if err != nil {
+			return nil, err
+		}
+
+		data, ok := templateList[c.template]
+		if !ok {
+			return nil, errors.New("template: " + c.template + " not found")
+		}
+		templateFile, err := writeTemplate(data)
+		if err != nil {
+			return nil, err
+		}
+		list = []string{
+			filepath.Join(templateFile.Name()),
+			filepath.Join(baseTemplate.Name()),
+		}
+	} else {
+		list = []string{
+			filepath.Join(c.TemplatePath, c.template),
+			filepath.Join(c.TemplatePath, "base.tmpl"),
+		}
+	}
+
+	var out *os.File
+	tmpl, err := template.ParseFiles(list...)
+	if err != nil {
+		return nil, err
+	}
+
+	if c.WriteFile {
+		if filepath.Ext(c.output) != ".html" {
+			c.output += ".html"
+		}
+		var createErr error
+		f, createErr := os.Create(filepath.Join(c.OutputPath, c.output))
+		if createErr != nil {
+			return nil, err
+		}
+		createErr = c.writeJavascriptLibrary()
+		if createErr != nil {
+			log.Errorf(log.Global, "failed to write javascript library: %v", createErr)
```

Comment:
This seems like it should return here too.

Reply:
Yeah I was originally undecided on if I should fail if just the chart library write fails since a user can manually add it
But I think i will go ahead and fail and clean up if it does